src/OpalCompiler-Tests/OCSpecialSelectorTest.class.st

```diff
@@ -0,0 +1,122 @@
+Class {
+	#name : #OCSpecialSelectorTest,
+	#superclass : #OCASTTranslatorTest,
+	#instVars : [
+		'optimisationsActive',
+		'original'
+	],
+	#category : #'OpalCompiler-Tests-AST'
+}
+
+{ #category : #'tests - simple' }
+OCSpecialSelectorTest >> compilationContext [
+
+	| context |
+	context := super compilationContext.
+
+	context parseOptions: {
+			(optimisationsActive
+				 ifTrue: [ #+ ]
+				 ifFalse: [ #- ]).
+			#optionOptimiseSpecialSends }.
+	^ context
+]
+
+{ #category : #'tests - simple' }
+OCSpecialSelectorTest >> proxy: aMethod [
+	"Replace the method by the proxy and keep the original one to restore it on tearDown"
+
+	| proxy |
+	original := aMethod.
+	proxy := OCCalledMethodProxy on: original.
+	aMethod methodClass
+		addSelectorSilently: aMethod selector
+		withMethod: proxy.
+	^ proxy
+]
+
+{ #category : #'tests - simple' }
+OCSpecialSelectorTest >> setUp [
+
+	super setUp.
+	optimisationsActive := true
+]
+
+{ #category : #'tests - simple' }
+OCSpecialSelectorTest >> tearDown [
+
+	original methodClass
+		addSelectorSilently: original selector
+		withMethod: original.
+	super tearDown
+]
+
+{ #category : #'tests - simple' }
+OCSpecialSelectorTest >> testOptimisedPlusSpecialSendsMessageDoesNotCaptureSend [
+	"Turn on optimisation, then compile & run"
+
+	| proxy |
+	optimisationsActive := true.
+	self compileExample: #exampleReturn1plus2.
+
+	[
+		proxy := self proxy: SmallInteger >> #+.
+
+		self deny: proxy wasCalled.
+		self assert: (self executeMethod: method onReceiver: instance) equals: 3.
+		self deny: proxy wasCalled.
+	] valueUnpreemptively
+]
+
+{ #category : #'tests - simple' }
+OCSpecialSelectorTest >> testOptimisedValueSpecialSendsMessageDoesNotCaptureSend [
+	"Turn on optimisation, then compile & run"
+
+	| proxy |
+	optimisationsActive := true.
+	self compileExample: #exampleSimpleBlockiVar.
+	instance iVar: #valueToTest.
+	[
+		instance class addSelectorSilently: #exampleSimpleBlockiVar withMethod: method.
+		proxy := self proxy: BlockClosure >> #value.
+
+		self deny: proxy wasCalled.
+		self assert: instance exampleSimpleBlockiVar equals: #valueToTest.
```

Comment:
Maybe it is worth to add a comment in here to say why we send a message instead of calling `(self executeMethod: method onReceiver: instance)`. In the dojo we saw that is because of the JIT compiler. Else will be difficult to understand the why

Reply:
Yes!